crates/telio-firewall/benches/firewall_bench.rs

```diff
@@ -1,40 +1,97 @@
-use std::fmt::Display;
+use std::fmt::{Debug, Display, Formatter};
+use std::net::SocketAddr;
+use std::ops::Deref;
+use std::rc::Rc;
 
 use criterion::{criterion_group, criterion_main, BenchmarkId, Criterion};
 use pnet_packet::tcp::TcpFlags;
 use telio_crypto::SecretKey;
-use telio_firewall::firewall::tests::{make_tcp, make_udp};
+use telio_firewall::firewall::tests::{make_tcp, make_tcp6, make_udp, make_udp6};
 use telio_firewall::firewall::{Firewall, StatefullFirewall};
 
 const PEER_COUNTS: [usize; 8] = [0, 1, 2, 4, 8, 16, 32, 64]; // max 60 peers: https://meshnet.nordvpn.com/getting-started/meshnet-explained#meshnet-scalability
-const PACKET_COUNTS: [u64; 2] = [1_000, 100_000];
+const PACKET_COUNT: u64 = 100_000;
 
-#[derive(Debug, Clone, Copy)]
+#[derive(Clone)]
+struct Packet {
+    src: SocketAddr,
+    dst: SocketAddr,
+    payload: Vec<u8>,
+    packet_maker: Rc<dyn Fn(&str, &str) -> Vec<u8>>,
+}
+
+impl Debug for Packet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Packet")
+            .field("src", &self.src)
+            .field("dst", &self.dst)
+            .field("payload", &self.payload)
+            .finish()
+    }
+}
+impl Deref for Packet {
+    type Target = [u8];
+
+    fn deref(&self) -> &Self::Target {
+        &self.payload
+    }
+}
+
+impl Packet {
+    /// Create new packet and precompute new serialized packet in `payload` field.
+    fn new(src: &str, dst: &str, packet_maker: impl Fn(&str, &str) -> Vec<u8> + 'static) -> Self {
+        let payload = packet_maker(src, dst);
+        Self {
+            src: src.parse().unwrap(),
+            dst: dst.parse().unwrap(),
+            payload,
+            packet_maker: Rc::new(packet_maker),
+        }
+    }
+
+    /// Create new packet based on the packet making function used when creating `self`.
+    fn make(&self, src: impl ToString, dst: impl ToString) -> Vec<u8> {
+        (self.packet_maker)(&src.to_string(), &dst.to_string())
+    }
+}
+
+#[derive(Debug, Clone)]
 struct Parameter {
     peers: usize,
-    packets: u64,
+    packet: Packet,
 }
 
 impl Display for Parameter {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        write!(f, "{self:?}")
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let ip = if self.packet.src.is_ipv4() { 4 } else { 6 };
+        write!(f, "peers: {}, ip: v{ip}", self.peers)
```

Comment:
I assume it cannot look like: `"peers: {self.peers}, ip: v{ip}"` because of some syntax limitaion?

Reply:
yes, only identifier are allowed: https://rust-lang.github.io/rfcs/2795-format-args-implicit-identifiers.html#future-possibilities